instance/nomad_client.py

```diff
@@ -0,0 +1,91 @@
+# -*- coding: utf-8 -*-
+#
+# OpenCraft -- tools to aid developing and hosting free software projects
+# Copyright (C) 2015-2018 OpenCraft <contact@opencraft.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Convenience wrappers for calls to the Nomad REST API.
+"""
+
+# Imports #####################################################################
+
+import functools
+
+from django.conf import settings
+from django.core.exceptions import ImproperlyConfigured
+import requests
+
+
+# Functions ###################################################################
+
+def request(method, *endpoint_parts, server=settings.NOMAD_SERVER, version="v1", **kwargs):
+    """Send a REST API request to the Nomad API and return the Response object."""
+    if server is None:
+        raise ImproperlyConfigured("Please set NOMAD_SERVER in the Django settings.")
+    url = "https://" + "/".join([server, version, *endpoint_parts])
+    kwargs.setdefault("verify", settings.NOMAD_CACERT)
+    kwargs.setdefault("cert", (settings.NOMAD_CLIENT_CERT, settings.NOMAD_CLIENT_KEY))
+    kwargs.setdefault("timeout", 15)
+    return requests.request(method, url, **kwargs)
+
+# Convenience wrappers for specific HTTP methods
+get = functools.partial(request, "get")
+post = functools.partial(request, "post")
+put = functools.partial(request, "put")
+delete = functools.partial(request, "delete")
+
+
+# Classes #####################################################################
+
+class NomadJob:
+    """Class to represent a Nomad job.
+
+    This representation is meant to be ephemeral, and it is not persisted to the database.
```

Comment:
Would it be costly to store this persistently? I feel like that'll give us a nice way to track leaked jobs, which could become a serious issue requiring a lot of manual work to set right otherwise. If a request to stop fails, for example, we can just set a flag on the job and come back to it later in a periodic Huey job. Overall, I think there's a lot to gain by making job metadata persistent.

Reply:
For a general resource management engine, you definitely need a representation of the expected state in the database.  Since we unfortunately don't have a resource management engine, we need to decide this on an ad-hoc basis and just implement enough to get the job done in each individual case.

Implementing a scheduled job to kill old Nomad jobs doesn't really need a representation of it's full configuration in the database.  We could just as well iterate over all archived instances in the DB, do `get_nomad_jobs()` on them and stop them, this should be perfectly fine for our use case, and it's a lot less effort.

Thinking about it, it's actually easy enough to include it in this PR, so I will.

Reply:
Thanks, I also hope that'll work out just fine!